File: frontend/types/scenarios.ts
export type ScenarioType = "baseline" | "improved" | "target_bacs" | "custom";
export type ScenarioStatus = "draft" | "ready" | "archived";
export type TargetScope = "project" | "zone" | "system";

export type ScenarioResponse = {
  id: string;
  project_id: string;
  name: string;
  description: string | null;
  scenario_type: ScenarioType;
  status: ScenarioStatus;
  derived_from_scenario_id: string | null;
  is_reference: boolean;
  created_at: string;
  updated_at: string;
};

export type ScenarioCreatePayload = {
  name: string;
  description: string | null;
  scenario_type: ScenarioType;
  derived_from_scenario_id: string | null;
  is_reference: boolean;
};

export type ScenarioUpdatePayload = Partial<{
  name: string;
  description: string | null;
  scenario_type: ScenarioType;
  status: ScenarioStatus;
  is_reference: boolean;
}>;

export type ScenarioDuplicatePayload = {
  name: string | null;
};

export type SolutionCatalogItem = {
  id: string;
  catalog_id: string;
  code: string;
  name: string;
  description: string;
  solution_family: string;
  family: string;
  target_scopes: TargetScope[];
  applicable_countries: string[];
  applicable_building_types: string[];
  applicable_zone_types: string[];
  bacs_impact_json: Record<string, unknown>;
  lifetime_years: number | null;
  default_quantity: number | null;
  default_unit: string | null;
  default_unit_cost: number | null;
  default_capex: number | null;
  priority: number;
  scope: string;
  country_code: string | null;
  is_commercial_offer: boolean;
  offer_reference: string | null;
  is_active: boolean;
};

export type ScenarioSolutionAssignment = {
  id: string;
  scenario_id: string;
  solution_code: string;
  solution_name: string;
  solution_description: string;
  solution_family: string;
  target_scope: TargetScope;
  target_zone_id: string | null;
  target_system_id: string | null;
  quantity: number | null;
  unit_cost_override: number | null;
  capex_override: number | null;
  maintenance_override: number | null;
  gain_override_percent: number | null;
  notes: string | null;
  is_selected: boolean;
  created_at: string;
  updated_at: string;
};

export type ScenarioSolutionAssignmentCreatePayload = {
  solution_code: string;
  target_scope: TargetScope;
  target_zone_id: string | null;
  target_system_id: string | null;
  quantity: number | null;
  unit_cost_override: number | null;
  capex_override: number | null;
  maintenance_override: number | null;
  gain_override_percent: number | null;
  notes: string | null;
  is_selected: boolean;
};

export type ScenarioSolutionAssignmentUpdatePayload = Partial<Omit<ScenarioSolutionAssignmentCreatePayload, "solution_code">>;

File: backend/app/schemas/scenarios.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

ScenarioType = Literal["baseline", "improved", "target_bacs", "custom"]
ScenarioStatus = Literal["draft", "ready", "archived"]
TargetScope = Literal["project", "zone", "system"]


class ScenarioCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    scenario_type: ScenarioType = "custom"
    derived_from_scenario_id: UUID | None = None
    is_reference: bool = False


class ScenarioUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    scenario_type: ScenarioType | None = None
    status: ScenarioStatus | None = None
    is_reference: bool | None = None


class ScenarioDuplicateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)


class ScenarioResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    name: str
    description: str | None
    scenario_type: ScenarioType
    status: ScenarioStatus
    derived_from_scenario_id: UUID | None
    is_reference: bool
    created_at: datetime
    updated_at: datetime


class SolutionCatalogItemResponse(BaseModel):
    code: str
    name: str
    description: str
    solution_family: str
    target_scopes: list[TargetScope]
    default_quantity: float | None = None
    default_unit: str | None = None


class ScenarioSolutionAssignmentCreate(BaseModel):
    solution_code: str
    target_scope: TargetScope = "project"
    target_zone_id: UUID | None = None
    target_system_id: UUID | None = None
    quantity: float | None = Field(default=None, gt=0)
    unit_cost_override: float | None = Field(default=None, ge=0)
    capex_override: float | None = Field(default=None, ge=0)
    maintenance_override: float | None = Field(default=None, ge=0)
    gain_override_percent: float | None = Field(default=None, ge=0, le=1)
    notes: str | None = None
    is_selected: bool = True

    @model_validator(mode="after")
    def validate_target(self) -> "ScenarioSolutionAssignmentCreate":
        if self.target_scope == "zone" and self.target_zone_id is None:
            raise ValueError("target_zone_id is required when target_scope is zone")
        if self.target_scope == "system" and self.target_system_id is None:
            raise ValueError("target_system_id is required when target_scope is system")
        return self


class ScenarioSolutionAssignmentUpdate(BaseModel):
    target_scope: TargetScope | None = None
    target_zone_id: UUID | None = None
    target_system_id: UUID | None = None
    quantity: float | None = Field(default=None, gt=0)
    unit_cost_override: float | None = Field(default=None, ge=0)
    capex_override: float | None = Field(default=None, ge=0)
    maintenance_override: float | None = Field(default=None, ge=0)
    gain_override_percent: float | None = Field(default=None, ge=0, le=1)
    notes: str | None = None
    is_selected: bool | None = None


class ScenarioSolutionAssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    scenario_id: UUID
    solution_code: str
    target_scope: TargetScope
    target_zone_id: UUID | None
    target_system_id: UUID | None
    quantity: float | None
    unit_cost_override: float | None
    capex_override: float | None
    maintenance_override: float | None
    gain_override_percent: float | None
    notes: str | None
    is_selected: bool
    created_at: datetime
    updated_at: datetime


class ScenarioSolutionDetailedResponse(ScenarioSolutionAssignmentResponse):
    solution_name: str
    solution_description: str
    solution_family: str
